src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -881,6 +1022,87 @@ func (s *commitLogSource) bootstrapShardBlockSnapshot(
 	return nil
 }
 
+func (s *commitLogSource) bootstrapIndexSnapshots(
+	ns namespace.Metadata,
+	indexResult result.IndexBootstrapResult,
+	mostRecentIndexSnapshotsByBlock map[xtime.UnixNano]indexSnapshot,
+) error {
+	for blockStart, snapshot := range mostRecentIndexSnapshotsByBlock {
+		if snapshot.fileSet.CachedSnapshotTime.Equal(blockStart.ToTime()) ||
+			// Should never happen
+			snapshot.fileSet.IsZero() {
+			// There is no snapshot file for this time, and even if there was, there would
+			// be no point in reading it. In this specific case its not an error scenario
+			// because the fact that snapshotTime == blockStart means we already accounted
+			// for the fact that this snapshot did not exist when we were deciding which
+			// commit logs to read.
+			s.log.Debug("no index snapshots for blockStart",
```

Comment:
Can you emit an invariant violated metric and log here using `instrument.EmitAndLogInvariantViolation`?

Reply:
I split the logic here since there were two conditions, but only one of which is a true invariant violation (zero value snapshot).

Reply:
```
if mostRecentCompleteSnapshotForShardBlock.IsZero() {
        iOpts := s.opts.CommitLogOptions().InstrumentOptions()
        instrument.EmitAndLogInvariantViolation(iOpts, func(l *zap.Logger) {
                l.Error(fmt.Sprintf("zero value snapshot for shard: %d blockStart: %v", shard, blockStart))
        })
}
if mostRecentCompleteSnapshotForShardBlock.CachedSnapshotTime.Equal(blockStart) {
        // There is no snapshot file for this time, and even if there was, there would
        // be no point in reading it. In this specific case its not an error scenario
        // because the fact that snapshotTime == blockStart means we already accounted
        // for the fact that this snapshot did not exist when we were deciding which
        // commit logs to read.
        s.log.Debug("no index snapshots for blockStart",
                zap.Uint32("shard", shard), zap.Time("blockStart", blockStart))
        continue
}
```

Reply:
Hmm actually... upon further digging, it seems like we always had a snapshot entry per block start regardless of whether we have a snapshot for that block start or not. We just default the `CachedSnapshotTime` to the block start if there's no snapshot.

I'm not sure what the original intention of the snapshot `IsZero()` check here was... but it seems like it should not be an invariant. My best guess is that we were intending on using the `CachedSnapshotTime` on a per block basis to determine from when to start reading commit logs from? But we have a global snapshot metadata file that has a commit log ID we can use to determine this info. 

Also, the only place I see `CachedSnapshotTime` used is to determine whether or not to skip loading a snapshot per (shard and) block start.